src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -403,29 +403,34 @@ bool ObjectMonitor::enter(JavaThread* current) {
     }
 
     OSThreadContendState osts(current->osthread());
-    ThreadBlockInVM tbivm(current);
 
-    // TODO-FIXME: change the following for(;;) loop to straight-line code.
-    for (;;) {
-      current->set_suspend_equivalent();
-      // cleared by handle_special_suspend_equivalent_condition()
-      // or java_suspend_self()
+    assert(current->thread_state() == _thread_in_vm, "invariant");
 
+    current->frame_anchor()->make_walkable(current);
+    OrderAccess::storestore();
+    for (;;) {
+      current->set_thread_state(_thread_blocked);
       EnterI(current);
-
-      if (!current->handle_special_suspend_equivalent_condition()) break;
-
-      // We have acquired the contended monitor, but while we were
-      // waiting another thread suspended us. We don't want to enter
-      // the monitor while suspended because that would surprise the
-      // thread that suspended us.
-      //
-      _recursions = 0;
-      _succ = NULL;
-      exit(current, false /* not_suspended */);
-
-      current->java_suspend_self();
+      current->set_thread_state_fence(_thread_blocked_trans);
+      if (SafepointMechanism::should_process(current) &&
+        current->suspend_request_pending()) {
+        // We have acquired the contended monitor, but while we were
+        // waiting another thread suspended us. We don't want to enter
+        // the monitor while suspended because that would surprise the
+        // thread that suspended us.
+        _recursions = 0;
+        _succ = NULL;
+        exit(current, false /* not_suspended */);
+        SafepointMechanism::process_if_requested(current);
+        // Since we are going to _thread_blocked we skip setting _thread_in_vm here.
+      } else {
+        // Only exit path from for loop
+        SafepointMechanism::process_if_requested(current);
```

Comment:
I think the current thread can suspend here in L428. This seems problematic as entering OM has been started but not completed.

- The current thread is set as owner in ObjectMonitor::_owner
- The thread state will still be JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER because of the JavaThreadBlockedOnMonitorEnterState in L389.
- Thread::_current_pending_monitor has not been reset to NULL.

Reply:
Yes, I think you are correct. I'll look over it.

Reply:
Today the ThreadBlockInVM tbivm(current); is inside scope of JavaThreadBlockedOnMonitorEnterState jtbmes(current, this);.
So this can happen today also.

If you are context switch just before current->set_current_pending_monitor(NULL);.
Someone suspends you and look at those states.

If you agree that the issue is preexisting I prefer handling that outside scope of this.

Reply:
> Today the ThreadBlockInVM tbivm(current); is inside scope of JavaThreadBlockedOnMonitorEnterState jtbmes(current, this);.
> So this can happen today also.
> 
> If you are context switch just before current->set_current_pending_monitor(NULL);.
> Someone suspends you and look at those states.
> 

You mean the JVMTI agent suspends the current thread and then observes that the
thread state first has the attribute JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER and
in a later call it does not have it anymore (~ThreadBlockInVM doesn't check for
suspend)? Yes that's problematic too.

With the new code we could remain suspended with
JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER. I think the OM would not be
reported as owned monitor but another thread could not lock it.

> If you agree that the issue is preexisting I prefer handling that outside scope of this.

I'm ok with that.

A simple solution could then be making use of ThreadBlockInVM. When returning
from EnterI in L413 we would set a rollback function in the HandshakeState which
can be called in HandshakeState::suspend_in_handshake() to exit the OM.

Reply:
I'm mean the state you describe will be seen on that line:
- The current thread is set as owner in ObjectMonitor::_owner
- The thread state will still be JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER because of the JavaThreadBlockedOnMonitorEnterState in L389.
- Thread::_current_pending_monitor has not been reset to NULL.

Will be seen while we are context while in blocked state before clearing the _current_pending_monitor.

Ok good.